### PR TITLE
Add Cosine LR Scheduler for Fine-Tuning

### PR DESCRIPTION
**Pull request by azahed98**
*Have you read the [Contributing Guidelines](https://github.com/togethercomputer/together/blob/main/CONTRIBUTING.md)?*
Yes

Issue # ENG-23270

Adds support for Cosine LR Scheduler in the fine-tuning api as well as simplifies future support of additional schedulers (https://github.com/togethercomputer/together-finetune/pull/1349)

### CODE REVIEW

File: src/together/types/finetune.py [hunk 1]
```diff
@@ -345,13 +345,53 @@ class FinetuneTrainingLimits(BaseModel):
...
+
 class FinetuneLRScheduler(BaseModel):
     lr_scheduler_type: str
-    lr_scheduler_args: FinetuneLinearLRSchedulerArgs | None = None
+    lr_scheduler_args: FinetuneLRSchedulerArgs | None = None
+
+    @field_validator("lr_scheduler_type")
+    @classmethod
+    def validate_scheduler_type(cls, v: str) -> str:
+        if v not in LRSchedulerTypeToArgs:
```
Comment (mryab) [inline src/together/types/finetune.py:374]: Better not to use single-letter variables

File: src/together/resources/finetune.py [hunk 2]
```diff
@@ -652,8 +674,10 @@ async def create(
             batch_size (int, optional): Batch size for fine-tuning. Defaults to max.
             learning_rate (float, optional): Learning rate multiplier to use for training
                 Defaults to 0.00001.
+            lr_scheduler_type (Literal["linear", "cosine"]): Learning rate scheduler type. Defaults to "linear".
             min_lr_ratio (float, optional): Min learning rate ratio of the initial learning rate for
                 the learning rate scheduler. Defaults to 0.0.
+            num_cycles (float, optional): Number of cycles for cosine learning rate scheduler. Defaults to 0.5.
```
Comment (mryab) [inline src/together/resources/finetune.py:680]:
```suggestion
            num_cycles (float, optional): Number of cycles for the cosine learning rate scheduler. Defaults to 0.5.
```

File: src/together/resources/finetune.py [hunk 3]
```diff
@@ -279,8 +295,10 @@ def create(
             batch_size (int or "max"): Batch size for fine-tuning. Defaults to max.
             learning_rate (float, optional): Learning rate multiplier to use for training
                 Defaults to 0.00001.
+            lr_scheduler_type (Literal["linear", "cosine"]): Learning rate scheduler type. Defaults to "linear".
             min_lr_ratio (float, optional): Min learning rate ratio of the initial learning rate for
                 the learning rate scheduler. Defaults to 0.0.
+            num_cycles (float, optional): Number of cycles for cosine learning rate scheduler. Defaults to 0.5.
```
Comment (mryab) [inline src/together/resources/finetune.py:301]:
```suggestion
            num_cycles (float, optional): Number of cycles for the cosine learning rate scheduler. Defaults to 0.5.
```

File: src/together/cli/api/finetune.py [hunk 4]
```diff
@@ -71,12 +71,24 @@ def fine_tuning(ctx: click.Context) -> None:
...
     "--min-lr-ratio",
     type=float,
     default=0.0,
     help="The ratio of the final learning rate to the peak learning rate",
 )
+@click.option(
+    "--num-cycles",
+    type=float,
+    default=0.5,
+    help="Number of cycles for cosine learning rate scheduler.",
```
Comment (mryab) [inline src/together/cli/api/finetune.py:90]:
```suggestion
    help="Number of cycles for the cosine learning rate scheduler.",
```

File: src/together/types/finetune.py [hunk 5]
```diff
@@ -345,13 +345,53 @@ class FinetuneTrainingLimits(BaseModel):
...
+        scheduler_type = str(info.data.get("lr_scheduler_type"))
 
+        if v is None:
+            return v
 
-class FinetuneLinearLRSchedulerArgs(BaseModel):
-    min_lr_ratio: float | None = 0.0
+        expected_type = LRSchedulerTypeToArgs[scheduler_type]
+        if not isinstance(v, expected_type):
+            raise ValueError(f"Expected {expected_type}, got {type(v)}")
```
Comment (mryab) [inline src/together/types/finetune.py:392]: Maybe this should be TypeError? Not sure if Pydantic has any conventions around that

File: src/together/resources/finetune.py [hunk 6]
```diff
@@ -129,10 +132,21 @@ def createFinetuneRequest(
...
 
-    lrScheduler = FinetuneLRScheduler(
-        lr_scheduler_type="linear",
-        lr_scheduler_args=FinetuneLinearLRSchedulerArgs(min_lr_ratio=min_lr_ratio),
-    )
+    if lr_scheduler_type == "cosine":
+        if num_cycles <= 0.0:
+            raise ValueError("Number of cycles should be greater than 0")
+
+        lrScheduler = FinetuneLRScheduler(
```
Comment (artek0chumak) [inline src/together/resources/finetune.py:139]: Why not do separate classes for each LR scheduler type? It removes the need for custom validator and makes code cleaner

File: src/together/cli/api/finetune.py (hunk 4 above)
Comment (VProv) [reply src/together/cli/api/finetune.py:90]: nit: Maybe also add what fractional values mean.

File: src/together/resources/finetune.py [hunk 7]
```diff
@@ -57,7 +58,9 @@ def createFinetuneRequest(
     n_checkpoints: int | None = 1,
     batch_size: int | Literal["max"] = "max",
     learning_rate: float | None = 0.00001,
+    lr_scheduler_type: Literal["linear", "cosine"] = "linear",
     min_lr_ratio: float = 0.0,
+    num_cycles: float = 0.5,
```
[the remark this answers is not in the document]
Comment (azahed98) [reply src/together/resources/finetune.py:63]: The backend expects the field in the request to be `num_cycles`, but in the scheduler args. Would just changing the CLI arg be ok?

File: src/together/resources/finetune.py [hunk 8]
```diff
@@ -129,10 +132,21 @@ def createFinetuneRequest(
             f"training_method must be one of {', '.join(AVAILABLE_TRAINING_METHODS)}"
         )
 
-    lrScheduler = FinetuneLRScheduler(
-        lr_scheduler_type="linear",
-        lr_scheduler_args=FinetuneLinearLRSchedulerArgs(min_lr_ratio=min_lr_ratio),
-    )
+    if lr_scheduler_type == "cosine":
+        if num_cycles <= 0.0:
```
[the remark this answers is not in the document]
Comment (azahed98) [reply src/together/resources/finetune.py:136]: Hard to know what a reasonable upperbound would be without knowing the number of steps afaik. I think it makes sense to follow the hf implementation and let the cosine alias if the user inputs something unreasonably large for their job.

File: src/together/resources/finetune.py (hunk 6 above)
Comment (azahed98) [reply src/together/resources/finetune.py:139]: Fair point, will do this!

File: src/together/types/finetune.py (hunk 5 above)
Comment (azahed98) [reply src/together/types/finetune.py]: Changed

File: src/together/types/finetune.py (hunk 1 above)
Comment (azahed98) [reply src/together/types/finetune.py]: Resolved.

File: src/together/cli/api/finetune.py (hunk 4 above)
Comment (azahed98) [reply src/together/cli/api/finetune.py]: Updated to "Number or fraction of cycles".

File: src/together/resources/finetune.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (azahed98) [reply src/together/resources/finetune.py]: Changed CLI arg to `scheduler_num_cycles`. Worth noting `FinetuneCosineLRSChedulerArgs` still has the field as `num_cycles`
